libcaf_io/caf/io/basp/instance.hpp

```diff
@@ -368,49 +357,35 @@ class instance {
             return false;
           }
         }
-        if (tcp_based) {
-          if (tbl_.lookup_direct(hdr.source_node)) {
+        auto new_node = (this_node() != hdr.source_node
+                         && !tbl_.lookup(hdr.source_node));
+        if (!new_node) {
+          if (tcp_based) {
             CAF_LOG_INFO("received second client handshake:"
                          << CAF_ARG(hdr.source_node));
             break;
           }
-          // add direct route to this node and remove any indirect entry
-          CAF_LOG_INFO("new direct connection:" << CAF_ARG(hdr.source_node));
-          tbl_.add_direct(hdl, hdr.source_node);
-          auto was_indirect = tbl_.erase_indirect(hdr.source_node);
-          callee_.learned_new_node_directly(hdr.source_node, was_indirect);
         } else {
-          auto new_node = (this_node() != hdr.source_node
-                          && !tbl_.lookup_direct(hdr.source_node));
-          if (new_node) {
-            // add direct route to this node and remove any indirect entry
-            CAF_LOG_INFO("new direct connection:" << CAF_ARG(hdr.source_node));
-            tbl_.add_direct(hdl, hdr.source_node);
-          }
+          // Add this node to our contacts.
+          CAF_LOG_INFO("new endpoint:" << CAF_ARG(hdr.source_node));
+          tbl_.add(hdl, hdr.source_node);
+          // TODO: Add status, addresses, ... ?
```

Comment:
Same question here. What additional state do we need now that's not in the routing table?